internal/controller/maintenancemode_controller.go

```diff
@@ -0,0 +1,188 @@
+		sc.Name = mm.Spec.TargetID
+		err := r.get(sc)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		clientName := sc.Spec.StorageClient
+		if mapClientNameToObj[clientName] == nil {
+			storageClient := &v1alpha1.StorageClient{}
+			storageClient.Name = clientName
+			err := r.get(storageClient)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+			mapClientNameToObj[clientName] = storageClient
+		}
+		if mapClientNameToObj[clientName].Status.InMaintenanceMode {
+			err := r.updateStatusCompletedForMM(mm)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	}
+
+	storageClients := &v1alpha1.StorageClientList{}
+	if err := r.list(storageClients); err != nil {
+		r.log.Error(err, "failed to list the Storage Clients")
+		return reconcile.Result{}, err
+	}
+
+	for i := range storageClients.Items {
+		storageClient := &storageClients.Items[i]
+		if mapClientNameToObj[storageClient.Name] != nil && !storageClient.Status.InMaintenanceMode {
+			if err := r.sendRequest(storageClient, true); err != nil {
+				return ctrl.Result{}, err
+			}
+		} else if mapClientNameToObj[storageClient.Name] == nil && storageClient.Status.InMaintenanceMode {
+			if err := r.sendRequest(storageClient, false); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *MaintenanceModeReconciler) sendRequest(storageClient *v1alpha1.StorageClient, enable bool) error {
+	providerClient, err := providerclient.NewProviderClient(
+		r.ctx,
+		storageClient.Spec.StorageProviderEndpoint,
+		10*time.Second,
+	)
+	if err != nil {
+		return fmt.Errorf(
+			"failed to create provider client with endpoint %v: %v",
+			storageClient.Spec.StorageProviderEndpoint,
+			err,
+		)
+	}
+	// Close client-side connections.
+	defer providerClient.Close()
+
+	_, err = providerClient.RequestMaintenanceMode(r.ctx, storageClient.Status.ConsumerID, enable)
+	if err != nil {
+		return fmt.Errorf("failed to Request maintenance mode: %v", err)
+	}
+	return nil
+}
+
+func (r *MaintenanceModeReconciler) updateStatusCompletedForMM(maintenanceMode *ramenv1alpha1.MaintenanceMode) error {
+	// Ramen reads the State and Conditions in order to determine that the MaintenanceMode is Completed
+	maintenanceMode.Status.State = ramenv1alpha1.MModeStateCompleted
+	maintenanceMode.Status.ObservedGeneration = maintenanceMode.Generation
+	meta.SetStatusCondition(&maintenanceMode.Status.Conditions,
+		metav1.Condition{
+			Type:               string(ramenv1alpha1.MModeConditionFailoverActivated),
+			ObservedGeneration: maintenanceMode.Generation,
+			Reason:             string(ramenv1alpha1.MModeStateCompleted),
+			Status:             metav1.ConditionTrue,
+		},
+	)
+
+	statusErr := r.Client.Status().Update(r.ctx, maintenanceMode)
+	if statusErr != nil {
+		r.log.Error(statusErr, "Failed to update MaintenanceMode status", "MaintenanceMode", maintenanceMode.Name)
+		return statusErr
+	}
```

Comment:
1. Simplify
2. Error logging should happen at the upper level
```suggestion
	if err := r.Client.Status().Update(r.ctx, maintenanceMode); err != nil {
		return err
	}
```